genpolicy: support named user mapping

HernanGatta opened the pull request:
Some containers specify their user via a name (e.g., Spark). As a result, instead of there being a UID:GID pair in the user field, there is instead a string (e.g., 'spark') that only resolves to a UID:GID pair at runtime.

Add a field to the `genpolicy` configuration file that allows specifying what UID:GID pair a given named user corresponds to. That way, the generated policy enforces the correct pair instead of rejecting the creation of the container, since the tool defaults to 0:0.

Fixes #9928

danmihai1 commented:
@HernanGatta , does https://github.com/kata-containers/kata-containers/issues/9928 sound like the same issue you are trying to address here?

If yes, I still hope that we'll be able to overlay the image layers and figure out the UID:GID from the overlay someday.

HernanGatta commented:
> @HernanGatta , does #9928 sound like the same issue you are trying to address here?
> 
> If yes, I still hope that we'll be able to overlay the image layers and figure out the UID:GID from the overlay someday.

@danmihai1 Yes, I included that issue in the commit's "Fixes" line.

I would indeed prefer to figure the UID:GID from the container image itself, though I'm not familiar enough (yet) with the tool's internals to understand how much work it would be to mount all the overlays, then go parse `/etc/passwd`.

danmihai1 commented:
> > @HernanGatta , does #9928 sound like the same issue you are trying to address here?
> > If yes, I still hope that we'll be able to overlay the image layers and figure out the UID:GID from the overlay someday.
> 
> @danmihai1 Yes, I included that issue in the commit's "Fixes" line.
> 
> I would indeed prefer to figure the UID:GID from the container image itself, though I'm not familiar enough (yet) with the tool's internals to understand how much work it would be to mount all the overlays, then go parse `/etc/passwd`.

The relevant code is around get_image_layers(). 

CC: @Redent0r , @burgerdev. Hopefully one of us can figure out a good way to overlay these files, and parse /etc/passwd.

burgerdev commented:
We would not even need to overlay them - it should be enough to find the last layer containing an `/etc/passwd` file (respecting whiteouts).

HernanGatta commented:
By way of update, I'm actively working on this. I've part of it working, and I'm currently debugging an issue that I presume is due to my unfamiliarity with Rust. I'll push an update once I figure it out.

danmihai1 commented:
> By way of update, I'm actively working on this. I've part of it working, and I'm currently debugging an issue that I presume is due to my unfamiliarity with Rust. I'll push an update once I figure it out.

That sounds great, thank you @HernanGatta !

HernanGatta committed:
genpolicy: pull UID:GID values from /etc/passwd

Some container images are configured such that the user (and group)
under which their entrypoint should run is not a number (or pair of
numbers), but a user name.

For example, in a Dockerfile, one might write:

> USER 185

indicating that the entrypoint should run under UID=185.

Some images, however, might have:

> RUN groupadd --system --gid=185 spark
> RUN useradd --system --uid=185 --gid=spark spark
> ...
> USER spark

indicating that the UID:GID pair should be resolved at runtime via
/etc/passwd.

To handle such images correctly, read through all /etc/passwd files in
all layers, find the latest version of it (i.e., the top-most layer with
such a file), and, in so doing, ensure that whiteouts of this file are
respected (i.e., if one layer adds the file and some subsequent layer
removes it, don't use it).

Signed-off-by: Hernan Gatta <hernan.gatta@opaque.co>

HernanGatta committed:
genpolicy: add tar dependency

Used to read /etc/passwd from tar files.

Signed-off-by: Hernan Gatta <hernan.gatta@opaque.co>

HernanGatta commented:
@danmihai1 I've updated the change to use `/etc/passwd` as requested.

With this change, the tool generates policies for our workloads that have the correct `uid` and `gid` values.

I have some thoughts about what I've done, and ways of doing things differently, but knowing that I tend to overcomplicate things, I'll let you be the judge of how horrible this is, and do please pardon my Rust.

HernanGatta commented on src/tools/genpolicy/src/registry.rs:
```diff
@@ -453,7 +557,22 @@ pub fn get_verity_hash_value(path: &Path) -> Result<String> {
+        } else if path_str == PASSWD_FILE_WHITEOUT_TAR_PATH {
+            passwd = WHITEOUT_MARKER.to_owned();
+        }
+        break;
```
Mistake in latest update, will fix.

burgerdev commented on src/tools/genpolicy/src/registry.rs:
```diff
@@ -133,28 +163,87 @@ impl Container {
+        /*
+         * The user field may:
+         *
+         * 1. Be empty
+         * 2. Contain only a UID
+         * 3. Contain a UID:GID pair, in that format
+         * 4. Contain a user name, which we need to translate into a UID/GID pair
```
This list is not exhaustive[^1]: we might also get a `username:gid`, `uid:groupname` etc. I think it might be better to split at `:` and then treat the first item as in the `else` block below (213ff.). If there is a second item, we can either leave the int parsing (203ff.) for now , or you could add similar functionality for `/etc/group`.



[^1]: https://github.com/opencontainers/image-spec/blob/8797c3fedb77200c102c816ddcd295bb1e19d3e3/config.md?plain=1#L140

danmihai1 replied:
@burgerdev @HernanGatta , I think it's OK to file a github issue for future improvements and make those improvements later.

burgerdev commented on src/tools/genpolicy/src/registry.rs:
```diff
@@ -133,28 +163,87 @@ impl Container {
+                                            records.iter().find(|&r| r.user == *image_user)
+                                        {
+                                            process.User.UID = record.uid;
+                                            process.User.GID = record.gid;
```
I just remembered the big warning sign here: https://docs.docker.com/reference/dockerfile/#user. It's not only correct to use the gid from the passwd file here, but I think we should even do that lookup if the field is a uid (without a gid).

burgerdev commented on src/tools/genpolicy/src/registry.rs:
```diff
@@ -453,7 +557,23 @@ pub fn get_verity_hash_value(path: &Path) -> Result<String> {
+
+    let mut passwd = String::new();
+    for entry_wrap in tar::Archive::new(file).entries()? {
+        let mut entry = entry_wrap?;
+        let entry_path = entry.header().path()?;
```
Do we really want to return early if these fail, or rather continue the loop?

danmihai1 replied:
This kind of error suggests to me that we're dealing with a corrupted tar file. So, returning early seems reasonable to me - at least until/unless we'll find examples that need workarounds.

burgerdev replied:
I checked the [docs](https://docs.rs/tar/latest/tar/struct.Header.html#method.path), should be fine to fail here:

> This method may fail if the pathname is not valid Unicode and this is called on a Windows platform.

burgerdev commented on src/tools/genpolicy/src/registry.rs:
```diff
@@ -453,7 +557,23 @@ pub fn get_verity_hash_value(path: &Path) -> Result<String> {
+        let mut entry = entry_wrap?;
+        let entry_path = entry.header().path()?;
+        let path_str = entry_path.to_str().unwrap();
+        if path_str == PASSWD_FILE_TAR_PATH {
+            entry.read_to_string(&mut passwd)?;
+            break;
+        } else if path_str == PASSWD_FILE_WHITEOUT_TAR_PATH {
+            passwd = WHITEOUT_MARKER.to_owned();
+            break;
+        }
```
Now I genuinely wonder whether the spec deals with both the file and a whiteout marker in the same layer...

burgerdev commented on src/tools/genpolicy/src/registry.rs:
```diff
@@ -64,8 +65,37 @@ pub struct DockerRootfs {
+struct PasswdRecord {
+    pub user: String,
+    #[allow(dead_code)]
+    pub validate: bool,
```
Nit: I believe the second field has three interpretations: login disabled, password login, passwordless. Doesn't matter much for this implementation, though.

burgerdev commented on src/tools/genpolicy/src/registry.rs:
```diff
@@ -508,3 +628,31 @@ fn build_auth(reference: &Reference) -> RegistryAuth {
+        records.push(PasswdRecord {
+            user: fields[0].to_string(),
+            validate: fields[1] == "x",
+            uid: fields[2].parse().unwrap(),
+            gid: fields[3].parse().unwrap(),
```
Did you mean to use the question-mark operator?

danmihai1 replied:
genpolicy panics on various paths when it deals with unexpected input format. So, the unwrap is consistent with the current design.

burgerdev commented on src/tools/genpolicy/src/registry.rs:
```diff
@@ -508,3 +628,31 @@ fn build_auth(reference: &Reference) -> RegistryAuth {
     RegistryAuth::Anonymous
 }
+
+fn parse_passwd_file(passwd: String) -> Result<Vec<PasswdRecord>> {
```
I just enabled tests for genpolicy the other day - how about adding one for this parser?

burgerdev reviewed (commented):
Thanks a lot, this is a nice addition and matches what I had in mind when thinking about the problem.

danmihai1 commented on src/tools/genpolicy/src/registry.rs:
```diff
@@ -133,28 +163,87 @@ impl Container {
         debug!("Getting process field from docker config layer...");
         let docker_config = &self.config_layer.config;
 
+        // Defaults to start with if we can't work out the correct values.
```
Unless you are seeing test failures, let's remove these defaults - because the intended design was that the caller provides the default values.

danmihai1 reviewed (commented):
@HernanGatta @burgerdev, the genpolicy code changes looks to me good enough to be merged. Thanks for the contribution, Hernan! I agree with @burgerdev that future improvements might be useful.

@HernanGatta, are you available to start looking at adding a simple test - maybe based on one of the images from https://quay.io/search?q=spark? A recent example of similar test: https://github.com/kata-containers/kata-containers/pull/10153.

If you are not available to work on that test soon, we could file an issue for it and either you or someone from MSFT can look at it a bit later.

We should also file issues as reminders to look into the other ideas for improvements from @burgerdev. 

@burgerdev, please speak up in case you think we need to delay the merge of this PR.
